src/storage/hummock_sdk/src/version.rs

```diff
@@ -0,0 +1,184 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::mem::size_of;
+
+use prost::Message;
+use risingwave_common::catalog::TableId;
+use risingwave_pb::hummock::hummock_version::PbLevels;
+use risingwave_pb::hummock::hummock_version_delta::PbGroupDeltas;
+use risingwave_pb::hummock::{PbHummockVersion, PbHummockVersionDelta};
+
+use crate::table_watermark::TableWatermarks;
+use crate::{CompactionGroupId, HummockSstableObjectId};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct HummockVersion {
+    pub id: u64,
+    pub levels: HashMap<CompactionGroupId, PbLevels>,
+    pub max_committed_epoch: u64,
+    pub safe_epoch: u64,
+    pub table_watermarks: HashMap<TableId, TableWatermarks>,
+}
+
+impl Default for HummockVersion {
+    fn default() -> Self {
+        HummockVersion::from_protobuf_inner(&PbHummockVersion::default())
+    }
+}
+
+impl HummockVersion {
+    pub fn from_rpc_protobuf(pb_version: &PbHummockVersion) -> Self {
+        Self::from_protobuf_inner(pb_version)
+    }
+
+    pub fn from_persisted_protobuf(pb_version: &PbHummockVersion) -> Self {
+        Self::from_protobuf_inner(pb_version)
+    }
+
+    fn from_protobuf_inner(pb_version: &PbHummockVersion) -> Self {
+        Self {
+            id: pb_version.id,
+            levels: pb_version
+                .levels
+                .iter()
+                .map(|(group_id, levels)| (*group_id as CompactionGroupId, levels.clone()))
+                .collect(),
+            max_committed_epoch: pb_version.max_committed_epoch,
+            safe_epoch: pb_version.safe_epoch,
+            table_watermarks: pb_version
+                .table_watermarks
+                .iter()
+                .map(|(table_id, table_watermark)| {
+                    (
+                        TableId::new(*table_id),
+                        TableWatermarks::from_protobuf(table_watermark),
+                    )
+                })
+                .collect(),
+        }
+    }
+
+    pub fn to_protobuf(&self) -> PbHummockVersion {
+        PbHummockVersion {
+            id: self.id,
+            levels: self
+                .levels
+                .iter()
+                .map(|(group_id, levels)| (*group_id as _, levels.clone()))
+                .collect(),
+            max_committed_epoch: self.max_committed_epoch,
+            safe_epoch: self.safe_epoch,
+            table_watermarks: self
+                .table_watermarks
+                .iter()
+                .map(|(table_id, watermark)| (table_id.table_id, watermark.to_protobuf()))
+                .collect(),
+        }
+    }
+
+    pub fn estimated_encode_len(&self) -> usize {
```

Comment:
Does the estimate of size need to correspond to all the member fields?

Reply:
This is just an estimated encode len. For simplicity I only include the encode len of fields with variable length (e.g. list and map).